### DISCUSSION TITLE
gis geometry node

### DISCUSSION THREAD

**Discussion by zeffii** [Marcus-Richmond/sverchok-gis-nodes#4] [Ideas]
I'm not so sure it is necessary to split points / lines and polygons into separate nodes. You could do that initially to get a grip on the data and potential duplication of effort. 

### Point node

there are more appropriate ways to loop these structures
```python
    # for loop to create vertices
    for index in range(len(gi['features'])):
            
        new_x = gi['features'][index]['geometry']['coordinates'][0]
        new_y = gi['features'][index]['geometry']['coordinates'][1]
        new_z = 0
        new_vector = [new_x, new_y, new_z]

        Vertices_features_individual.append(new_vector)
```
becomes... i hope...
```python
    ...
    for feature in gi['features']:
        x, y = feature['geometry']['coordinates']
        Vertices_features_individual.append([x, y, 0.0])
```
the goal is of course never to write as short as possible, but sometimes using python-idioms results in compacter code.

- I imagine those 3rd party libraries offers faster ways to get `np_array` from the gis data , if not we will make them. 
- do you know if polygons are stored as ngons, or are they pre-tessellated into triangles ?

**Comment by Marcus-Richmond** [comment]
Cool, I'm going to experiment with this a bit and try to get a solid geometry loader node put together. Knowing different structures like this is also incredibly helpful, so if you see any other improvements like this I'd be glad to hear about them! Thanks,

As far as the polygons - they are stored as ngons. The vertices for each polygon are also stored as coordinates, not as indices. For example, here is the geointerface method called on a test polygons layer:

```
import geopandas as gpd
import pandas as pd
import fiona
import numpy as np

# read in layer from gpkg
gpd1 = gpd.read_file("D:/0_RMaps/Projects/2400_BlenderAndGIS/Vector_Geopackage/BlenderAndGIS.gpkg", 
                    layer = "test shapes - polygons 1")

# call geointerface on gdf
gi = gpd1.__geo_interface__

gi
```
returns:
```
{'type': 'FeatureCollection',
 'features': [{'id': '0',
   'type': 'Feature',
   'properties': {'Decimal': 4.5, 'Integer': 2, 'Label': 'Label 1'},
   'geometry': {'type': 'Polygon',
    'coordinates': (((1.0211338730723973, 5.142472602082384),
      (3.727107447530596, 4.547382977823942),
      (5.590973063132509, 1.830181297247659),
      (8.611333608897052, 0.572633412022272),
      (9.161510808683161, 5.771246544695077),
      (2.705349790784968, 7.2758127645183075),
      (-0.6069415140497565, 5.69264980186849),
      (1.0211338730723973, 5.142472602082384)),)},
   'bbox': (-0.6069415140497565,
    0.572633412022272,
    9.161510808683161,
    7.2758127645183075)},
  {'id': '1',
   'type': 'Feature',
   'properties': {'Decimal': 8.99, 'Integer': 5, 'Label': 'Label 2'},
   'geometry': {'type': 'Polygon',
    'coordinates': (((-9.230127012738121, -2.2793055419710146),
      (0.9425371302458121, -5.221069344908974),
      (6.792380417767477, -2.3579022847976017),
      (9.36361671880867, -7.781077539832081),
      (-1.6286991707953806, -9.218275122946812),
      (-9.510829665690217, -4.3003646432261),
      (-9.230127012738121, -2.2793055419710146)),
     ((4.805005634866642, -4.985279116429215),
      (3.1207897171540715, -6.759319883086457),
      (6.567818295405804, -7.017566323802383),
      (4.805005634866642, -4.985279116429215)))},
   'bbox': (-9.510829665690217,
    -9.218275122946812,
    9.36361671880867,
    -2.2793055419710146)}],
 'bbox': (-9.510829665690217,
  -9.218275122946812,
  9.36361671880867,
  7.2758127645183075)}
```
Plus, this layer does not have z-values enabled, so you can see the coordinate vectors are just x/y. GIS data does not use z-values as much as Blender, so layers will often lack z-values.

**Comment by zeffii** [comment]
you might have noticed that features are stored in nested tuples, (sverchok does something similar by nesting lists) we will assume (for now) this is representative of all following features of this type. 

it would be ,
```python
feature_verts = []

gi = {'type': 'FeatureCollection',
 'features': [{'id': '0',
   'type': 'Feature',
   'properties': {'Decimal': 4.5, 'Integer': 2, 'Label': 'Label 1'},
   'geometry': {'type': 'Polygon',
    'coordinates': (((1.0211338730723973, 5.142472602082384),
      (3.727107447530596, 4.547382977823942),
      (5.590973063132509, 1.830181297247659),
      (8.611333608897052, 0.572633412022272),
      (9.161510808683161, 5.771246544695077),
      (2.705349790784968, 7.2758127645183075),
      (-0.6069415140497565, 5.69264980186849),
      (1.0211338730723973, 5.142472602082384)),)},
   'bbox': (-0.6069415140497565,
    0.572633412022272,
    9.161510808683161,
    7.2758127645183075)},
  {'id': '1',
   'type': 'Feature',
   'properties': {'Decimal': 8.99, 'Integer': 5, 'Label': 'Label 2'},
   'geometry': {'type': 'Polygon',
    'coordinates': (((-9.230127012738121, -2.2793055419710146),
      (0.9425371302458121, -5.221069344908974),
      (6.792380417767477, -2.3579022847976017),
      (9.36361671880867, -7.781077539832081),
      (-1.6286991707953806, -9.218275122946812),
      (-9.510829665690217, -4.3003646432261),
      (-9.230127012738121, -2.2793055419710146)),
     ((4.805005634866642, -4.985279116429215),
      (3.1207897171540715, -6.759319883086457),
      (6.567818295405804, -7.017566323802383),
      (4.805005634866642, -4.985279116429215)))},
   'bbox': (-9.510829665690217,
    -9.218275122946812,
    9.36361671880867,
    -2.2793055419710146)}],
 'bbox': (-9.510829665690217,
  -9.218275122946812,
  9.36361671880867,
  7.2758127645183075)}

for feature in gi['features']:
    nested_array_2d = feature['geometry']['coordinates']
    feature_verts.append( [[[c[0], c[1], 0.0] for c in seg] for seg in nested_array_2d])

print(feature_verts)
```
the `numpy` variety might prove slower when the amount of verts is relatively small.

**Reply by Marcus-Richmond** [reply]
Is there a benefit to using numpy arrays? Or, what is the benefit to using numpy arrays in Sverchok?

I see there is a ton of material online on numpy vs regular arrays, but i'm curious how these differences play out in Sverchok.

**Reply by zeffii** [reply]
as the size of the array becomes larger, numpy's benefits (speed, code elegance) is worth the extra overhead it introduces when converting existing data to numpy arrays..

the most intensively used Sverchok nodes (scalar/vec math.. etc) are capable of operating directly on numpy arrays. The more of the pipeline is perfomed as numpy arrays the faster the total will be. (with interesting exceptions which we may not encounter!)

**Reply by Marcus-Richmond** [reply]
Gotcha, sounds like we should use numpy arrays then (which I'm good with). Even while I've been testing the nodes I've used some fairly large datasets (bad habit for testing, I know ha) and would be glad to have the whole pipeline sped up a bit.
